akka-stream/src/main/scala/akka/stream/impl/fusing/MapAsyncPartitioned.scala

```diff
@@ -0,0 +1,237 @@
+    private var cachedSupervisionDirective: OptionVal[Supervision.Directive] = OptionVal.None
+
+    def supervisionDirectiveFor(decider: Supervision.Decider, ex: Throwable): Supervision.Directive =
+      cachedSupervisionDirective match {
+        case OptionVal.Some(d) => d
+        case _ =>
+          val d = decider(ex)
+          cachedSupervisionDirective = OptionVal.Some(d)
+          d
+      }
+
+    def clearIncoming(): Unit = {
+      incoming = null.asInstanceOf[I]
+    }
+
+    def setOutgoing(t: Try[O]): Unit = {
+      outgoing = t
+    }
+
+    override def apply(t: Try[O]): Unit = {
+      setOutgoing(t)
+      cb.invoke(this)
+    }
+
+    override def toString: String = s"Holder($incoming, $outgoing, $partition)"
+  }
+
+  val NotYetThere = MapAsync.NotYetThere
+}
+
+private[akka] sealed abstract case class MapAsyncPartitioned[In, Out, Partition] private (
+    parallelism: Int,
+    perPartition: Int,
+    partitioner: In => Partition,
+    f: (In, Partition) => Future[Out])
+    extends GraphStage[FlowShape[In, Out]] {
+  import MapAsyncPartitioned._
+
+  private val in = Inlet[In]("MapAsyncPartitioned.in")
+  private val out = Outlet[Out]("MapAsyncPartitioned.out")
+
+  override def initialAttributes = DefaultAttributes.mapAsync and SourceLocation.forLambda(f)
+  override val shape = FlowShape(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) with InHandler with OutHandler {
+
+      lazy val decider = inheritedAttributes.mandatoryAttribute[SupervisionStrategy].decider
+      var buffer: PartitionedBuffer[Partition, Holder[Partition, In, Out]] = _
+
+      private val futureCB = getAsyncCallback[Holder[Partition, In, Out]] { holder =>
+        holder.outgoing match {
+          case Failure(ex) if holder.supervisionDirectiveFor(decider, ex) == Supervision.Stop => failStage(ex)
+          case _ =>
+            dropCompletedThenPushIfPossible(holder.partition)
+        }
+      }
+
+      override def preStart(): Unit = {
+        buffer = new PartitionedBuffer[Partition, Holder[Partition, In, Out]](
+          linearBuffer = BufferImpl(parallelism, inheritedAttributes),
+          partitioner = _.partition,
+          makePartitionBuffer = { (p, overflowCapacity) =>
+            import akka.stream.impl.{ BoundedBuffer, ChainedBuffer }
+            val tailCapacity = (overflowCapacity - perPartition).max(1)
+
+            new ChainedBuffer[Holder[Partition, In, Out]](
+              BufferImpl(perPartition, inheritedAttributes),
+              new BoundedBuffer.DynamicQueue(tailCapacity), {
+                holder =>
+                  // this will execute when the holder moves from the tail buffer to the head buffer (viz. ready to be
+                  //  scheduled).  This might be on enqueueing into the partitioned buffer, or it might be when
+                  //  the head of this partition is dropped after being completed.
```

Comment:
Gut feeling is that this would be easier to follow if it was more imperative/part of the logic where it happens than as a callback handed to and executed by the buffer.

Reply:
To clarify: an object (the buffer) that changes itself under some circumstances and in this case also calls methods on the owning object vs a more collection like API where changes has to be explicitly decided by the callee so you know exactly when/where they are triggered.

Reply:
I strongly suspect that that approach would not be easier to follow in the aggregate because of the state that would be moved out of the buffers into the stage.

Reply:
That said, I wonder if it's possible to model the partitions inside the stage as something very stream like...